src/Gaffer/Animation.cpp

```diff
@@ -38,93 +38,1892 @@
+			const Imath::V2d mf( ( r2.x - time ) / ( 1.0 - time ), r2.y );
+			const Imath::V2d hi( ( r3.x - time ) / ( 1.0 - time ), r3.y );
+
+			newKey.setValue( l4.y );
+			newKey.getTangent( Gaffer::Animation::Tangent::Direction::Into ).setPosition( mi, false, space );
+			newKey.getTangent( Gaffer::Animation::Tangent::Direction::From ).setPosition( mf, false, space );
+			newTangentLo.setPosition( lf, false, space );
+			newTangentHi.setPosition( hi, false, space );
+		}
+
+	private:
+
+		double solveForTime( const double tl, const double th, const double time ) const
+		{
+			if( time <= 0.0 ) return 0.0;
+			if( time >= 1.0 ) return 1.0;
+
+			// compute coeffs
+
+			const double th3 = th + th + th;
+			const double ct = tl + tl + tl;
+			const double at = ct - th3 + 1.0;
+			const double bt = th3 - ct - ct;
+			const double bt2 = bt + bt;
+			const double at3 = at + at + at;
+
+			// root bracketed in interval [0,1]
+
+			double sl = 0.0;
+			double sh = 1.0;
+
+			// time is a reasonable first guess
+
+			double s = time;
+
+			// max of 10 newton-raphson iterations
+
+			for( int i = 0; i < 10; ++i )
+			{
+				// evaluate function and derivative
+				//
+				// NOTE : f   =  a(t)s^3 +  b(t)s^2 + c(t)s + d(t) - t
+				//        f'  = 3a(t)s^2 + 2b(t)s   + c(t)
+
+				const double  f = std::fma( s, std::fma( s, std::fma( s, at,  bt  ), ct ), -time );
+				const double df =              std::fma( s, std::fma( s, at3, bt2 ), ct );
+
+				// maintain bounds
+
+				if( std::abs( f ) < std::numeric_limits< double >::epsilon() )
+				{
+					break;
+				}
+				else if( f < 0.0 )
+				{
+					sl = s;
+				}
+				else
+				{
+					sh = s;
+				}
+
+				// NOTE : when derivative is zero or newton-raphson step would escape bounds use bisection step instead.
+
+				double ds;
+
+				if( df == 0.0 )
+				{
+					ds = 0.5 * ( sh - sl );
+					s = sl + ds;
+				}
+				else
+				{
+					ds = f / df;
+
+					if( ( ( s - ds ) <= sl ) || ( ( s - ds ) >= sh ) )
+					{
+						ds = 0.5 * ( sh - sl );
+						s = sl + ds;
+					}
+					else
+					{
+						s -= ds;
+					}
+				}
+
+				assert( s >= sl );
+				assert( s <= sh );
+
+				// check for convergence
+
+				if( std::abs( ds ) < std::numeric_limits< double >::epsilon() )
+				{
+					break;
+				}
+			}
+
+			return s;
+		}
+	};
+
+} // namespace
+
+namespace Gaffer
+{
+
+//////////////////////////////////////////////////////////////////////////
+// Time implementation
+//////////////////////////////////////////////////////////////////////////
+
+Animation::Time::Time()
+: m_ticks( static_cast< std::int64_t >( 0 ) )
+{}
+
+Animation::Time::Time( const std::int64_t ticks )
+: m_ticks( ticks )
+{}
+
+Animation::Time::Time( const double value, const Animation::Time::Units units )
+: m_ticks( static_cast< std::int64_t >( std::round( value * (
+	static_cast< double >( Units::Ticks ) /
+	static_cast< double >( units ) ) ) ) )
+{}
+
+Animation::Time::Time( const double value, const double units )
```

Comment:
Its less code for client code to write, we control the conversion and how its implemented rather than relying on client code. eg. if value is a float, client code may do value / units at float precision rather than double precision. Its main purpose is to interface with the existing context representation of time as a float in seconds. The existing context framespersecond is a float so I assume it can be non integer (otherwise why is it a float?). The min is because ticks is the max resolution supported. I cannot imagine anyone having a frame rate of nanoseconds.

Reply:
the new Animation::Time class has been removed from this PR and the code reverted to use float resolution time in seconds